### PR TITLE
Merge request for timer callback, UART and fast UART

### PR DESCRIPTION
**Pull request by RacingTornado**
Features:
- Callback timer functionality has been added. Upto 5 tasks can request for a timer. Currently I have modified the code so that only 1 callback is enabled in the ISR, but I can extend this functionality if you so desire . Adding features require a lot of testing. So the above library has been tested. The timer_lib.c functions and the ISR on timer 0 calculates this value. CUrrent the tick is set manually, although features for auto load and calculation have been provided in the library.  A 1ms callback appearing on PA3 can be observed on the FX2 pin
- Fast UART functionality has been added.  2 different types have been tested with a PL2303. One implementation takes 12 cycles, the second implementation takes 6 cycles. Precalculated delays have been calculated for a baud rate of 115200. This can be observed on PA2 of the FX2. The same fast_uart function also contains the second algorithm which takes about 12 cycles using a different algorithm. The details are located at soft_uart.c
- Timer based UART has been optimized and now operated on a baud of 9600. The timer 0 operates at 26us, which is 4x lesser than the period of a bit at 104us. Here the architecture is kind of critical. To enable the user to write and read from UART in the main while() loop , a Queue system has been introduced. You can push data onto the queue or read data from the queue. The pins PA4 and PA5 are the tx and rx pins for this UART mode. This operates at 9600 as already mentioned.
- There is also some part of the work which is in progress such as the mpsse_vendor.c which is for parsing stuff(I really have no clue what these commands I see on wireshark mean, so I have just ackd them back). Due to shortage of time, the USB to UART interface which was originally working has not been tested and is not part of this release. It will be looked at in the next commit.
- i2c_rw.c and spi_rw.c is still work in progress, the i2c_rw.c was working but the spi_rw.c has not been tested. Both should be considered libraries whose work is still in progress and not complete.

Some parts of this library such as the timer_lib.c require the caller to know what functions to set up. While this may not be the desirable way of achieving the required functionality, it simplifies things for a first try.

### CODE REVIEW

File: examples/i2c_mpsse/uart_serial.hex [hunk 1]
```diff
@@ -0,0 +1,310 @@
+:0400000002005B326D
```
Comment (mithro) [inline examples/i2c_mpsse/uart_serial.hex]: Don't commit generated files.

File: examples/i2c_mpsse/Makefile [hunk 2]
```diff
@@ -0,0 +1,124 @@
+#
+# Copyright (C) 2009-2012 Chris McClelland
+# Copyright 2015 Joel Stanley <joel@jms.id.au>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
```
Comment (mithro) [inline examples/i2c_mpsse/Makefile:7]: Please replace this Makefile with something closer to the other examples use.

File: examples/i2c_mpsse/bulkloop.c [hunk 3]
```diff
@@ -0,0 +1,438 @@
+/**
```
Comment (mithro) [inline examples/i2c_mpsse/bulkloop.c]: This file shouldn't be called bulkloop.c anymore

File: examples/i2c_mpsse/bulkloop.c [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+extern __xdata unsigned short periodic;
+extern unsigned char volatile tx_buffer;
+extern unsigned char volatile rx_buffer;
+extern unsigned char volatile tx_count;
+extern unsigned char volatile rx_count;
+extern unsigned char volatile tx_bits_sent;
+extern unsigned char volatile rx_bits_rcvd;
+extern unsigned char volatile tx_busy;
+extern unsigned char volatile rx_busy;
+extern void QueueInitTX(void);
```
Comment (mithro) [inline examples/i2c_mpsse/bulkloop.c]: Why is the formatting of these function names totally different to the ones above?

File: examples/i2c_mpsse/uart_config.c [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+         putchar_a(*src);
+      }
+   }
+
+   EP1OUTBC = 0xff; // re-arm endpoint 1 for OUT (host->device) transfers
+   SYNCDELAY;
+}
+
+
+void toggle_pins()
```
Comment (mithro) [inline examples/i2c_mpsse/uart_config.c:316]: This function seems to do nothing.

File: examples/i2c_mpsse/timer_lib.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+
+#define MAX_TIMERS 5
+
+#define CKCON_T0 0x08
+#define CKCON_T1 0x10
+#define CKCON_T2 0x20
+#define T0_MODE_REL 0x02
+
+#define ENABLE_INTERRUPT()  EA = 1
```
Comment (mithro) [inline examples/i2c_mpsse/timer_lib.h]: These style of macros already exist in fx2lib, you should use them.

File: examples/i2c_mpsse/toggle.c [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+#include <fx2ints.h>
+#include <autovector.h>
+#include <delay.h>
+#include <setupdat.h>
+
+
+#define SYNCDELAY SYNCDELAY4
+
+
+void configure_drive(unsigned char a, unsigned char b)
```
Comment (mithro) [inline examples/i2c_mpsse/toggle.c]: All these functions need comments.

File: examples/i2c_mpsse/timer_lib.c [hunk 8]
```diff
@@ -0,0 +1,175 @@
+/*
+
+Most of the logic here is implemented from this http://www.simplyembedded.org/tutorials/timers/
```
Comment (mithro) [inline examples/i2c_mpsse/timer_lib.c:3]: I don't see any license information on that blog you linked. If you have based your code on their implementation then you need to understand what license their code is under and if you can actually use it. You can't just copy and paste random code from the internet.

File: examples/i2c_mpsse/toggle.c [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+   }
+}
+
+
+
+void set_port_value(unsigned char a, unsigned char b)
+{
+   switch (a)
+   {
+      case 0x80:
```
Comment (mithro) [inline examples/i2c_mpsse/toggle.c]: Formatting.

File: examples/i2c_mpsse/toggle.c [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+      case 0xb1:
+            OEE |= b;
+         SYNCDELAY;
+         break;
+   }
+}
+
+
+
+void set_port_value(unsigned char a, unsigned char b)
```
Comment (mithro) [inline examples/i2c_mpsse/toggle.c]: I have NFI what this function is doing or why it is useful.

File: examples/i2c_mpsse/toggle.c [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+{
+   switch (a)
+   {
+      case 0x80:
+            IOA |= b;
+         SYNCDELAY;
+         break;
+
+      case 0x90:
```
Comment (mithro) [inline examples/i2c_mpsse/toggle.c]: What are these magic numbers in the case statement?

File: examples/i2c_mpsse/uart_config.c [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+
+/*
+ * USB setup packet descriptors
+ *                           ____________________________________________________________
+ *                          | Byte  | Field            | Meaning                          |
+ *                          |_______|__________________|__________________________________|
+                            | 0     | RequestType      |   Direction, and Recepient       |
+ *                          | 1     | Request          |   The actaul request             |
+ *                          | 2     | wValueL          |   16 bit value                   |
+ *                          | 3     | wValueH          |   16 bit value                   |
```
Comment (mithro) [inline examples/i2c_mpsse/uart_config.c:84]: Formatting..

Why is the table hugely indented?
Why do some lines start with \* but not others?

File: examples/i2c_mpsse/uart_config.c [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+
+unsigned int bytes_waiting_for_xmit = 0;
+unsigned int latency_us = 40000;
+//There are different endpoints on the FX2. However the CPU can access the endpoint 0 or endpoint 1 which is
+//smaller. In this case, the endpoint needs to be configured by setting the EPIOUTCFG, and EP1INCFG
+//registers . For endpoint configuration, refer Table 8.2
+
+
+
+void ProcessXmitData(void)
```
Comment (mithro) [inline examples/i2c_mpsse/uart_config.c:35]: Again you are switching back and forth between two styles of naming functions....

File: examples/i2c_mpsse/uart_config.c [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+#define FTDI_RS_FE (1<<3)
+#define FTDI_RS_BI (1<<4)
+#define FTDI_RS_THRE (1<<5)
+#define FTDI_RS_TEMT (1<<6)
+#define FTDI_RS_FIFO  (1<<7)
+
+
+unsigned int bytes_waiting_for_xmit = 0;
+unsigned int latency_us = 40000;
+//There are different endpoints on the FX2. However the CPU can access the endpoint 0 or endpoint 1 which is
```
Comment (mithro) [inline examples/i2c_mpsse/uart_config.c:29]: This comment needs work, it doesn't actually tell you why this is important in your use case.

File: examples/i2c_mpsse/uart_config.c [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+   ║         │ OUT/12 (for compatibility with standard 8051); when T1M = 1, ║
+   ║         │ Timer 1 uses CLKOUT/4.                                       ║
+   ╟─────────┼──────────────────────────────────────────────────────────────╢
+   ║ CKCON.3 │ T0M                                                          ║
+   ║         │ - Timer 0 clock select. When T0M = 0, Timer 0 uses CLK-      ║
+   ║         │ OUT/12 (for compatibility with standard 8051); when T0M = 1, ║
+   ║         │ Timer 0 uses CLKOUT/4                                        ║
+   ╚═════════╧══════════════════════════════════════════════════════════════╝
+   */
+   CKCON &= ~(0x08);
```
Comment (mithro) [inline examples/i2c_mpsse/uart_config.c:204]: Copying these tables and setting like this isnt good idea. You should have properly named defines and use them here.

**Comment by mithro** [comment]
Bunch of comments;
- Please split this pull request into parts. Each pull request should address one piece of functionality with examples. Generally you want a pull request to be _as small as possible_. The smaller it is, the better. IE One pull request for the fast UART, one for the timer based UART, one for I2C stuff, one for the SPI stuff.
- It is unclear what is actually working in this pull request. Only include working stuff in your pull request.
- At the moment this pull request is too confusing for me to tell if you have achieve the UART functionality we are after or not.
- There are no instructions on how to test this code and verify it works.

See my more specific comments on the code too.

File: examples/i2c_mpsse/timer_lib.c (hunk 8 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/timer_lib.c:3]: Scroll down on the ABOUT page. A lot of modfications have been made. It isnt just copy and paste, especially the timer critical section

File: examples/i2c_mpsse/uart_serial.hex (hunk 1 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/uart_serial.hex]: Will add this to my .gitignore

File: examples/i2c_mpsse/bulkloop.c (hunk 3 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/bulkloop.c]: Changed this

File: examples/i2c_mpsse/uart_config.c (hunk 5 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/uart_config.c:316]: I kept it for the USB functionality (send and receive on endpoint 1). Will remove it

File: examples/i2c_mpsse/timer_lib.h (hunk 6 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/timer_lib.h]: Understood. Changed it. I dont see any for global EA register, which is why I created them. Have updated it and directly used EA

File: examples/i2c_mpsse/toggle.c (hunk 7 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/toggle.c]: At this point of time I will remove these functions. They are only for debugging purposes

File: examples/i2c_mpsse/toggle.c (hunk 9 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/toggle.c]: Completely removed this file

File: examples/i2c_mpsse/toggle.c (hunk 10 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/toggle.c]: Removed this file.

File: examples/i2c_mpsse/toggle.c (hunk 11 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/toggle.c]: Removed the same file again

File: examples/i2c_mpsse/uart_config.c (hunk 12 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/uart_config.c:84]: Fixed

File: examples/i2c_mpsse/uart_config.c (hunk 13 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/uart_config.c:35]: I am not really sure what you mean here

File: examples/i2c_mpsse/uart_config.c (hunk 14 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/uart_config.c:29]: Updated why we use endpoint 1

File: examples/i2c_mpsse/Makefile (hunk 2 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/Makefile:7]: This has now been updated

File: examples/i2c_mpsse/bulkloop.c (hunk 4 above)
Comment (RacingTornado) [inline examples/i2c_mpsse/bulkloop.c]: I am still unclear about what you mean by this comment

**Pull request closed without merging**